### PR TITLE
Extend key expiry

### PR DESCRIPTION
**Pull request by bysnack**
This PR implements the extend  key expiry tool, it also intends to organize the repository a little bit by introducing some sort of hierarchy.

For the reviewers:
I tried to put every relevant change in a separated commit so it's easier to look at the changes, I recommend reviewing commit per commit.

### CODE REVIEW

File: src/extend_key_expiry/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+    options.cpp
+    ../shared/parameters.cpp
+    ../shared/parameters_ecdsa.cpp
+    ../shared/parameters_eddsa.cpp
+    ../shared/parameters_rsa.cpp
+    ../shared/packet_utils.cpp
+    ../shared/time_utils.cpp
+    ../shared/assert_release.cpp
+    ../shared/deterministic_rng.cpp
+    ../shared/util/options.cpp
```
Comment (bertptrs) [inline src/extend_key_expiry/CMakeLists.txt:13]: Does it make sense to compile the shared libraries as a static lib used by both rather than compile it twice?

File: src/extend_key_expiry/options.cpp [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+
+            // exit, since the user just requested help
+            exit(0);
+        }
+
+        // ensure that all the options are initialized by possibly reading some from standard input
+        input_file          .ensure_prompt("Public key file");
+        output_file         .ensure_prompt("Output file");
+        kdf_context         .ensure_prompt("Key derivation context (8 bytes)");
+        extension_period    .ensure_prompt("Key expiry extension period in months (3/6/9/12)");
```
Comment (bertptrs) [inline src/extend_key_expiry/options.cpp:63]: Why this overzealous whitespace?

File: src/extend_key_expiry/options.cpp [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+
+        // check that the KDF context is the right size
+        if (kdf_context->size() != 8) {
+            throw std::invalid_argument{ "Invalid key derivation context size, expected 8 bytes." };
+        }
+
+        // corroborate extension period
+        if (*extension_period != 3 && *extension_period != 6 && *extension_period != 12) {
```
Comment (bertptrs) [inline src/extend_key_expiry/options.cpp:71]: Doesn't this potentially dereference an empty optional?

File: src/extend_key_expiry/public_key.cpp [hunk 4]
```diff
@@ -0,0 +1,191 @@
+#include "public_key.h"
+
+#include "parameters_eddsa.h"
+#include "parameters_ecdsa.h"
+#include "parameters_rsa.h"
+#include "generate_key.h"
+
+
+namespace {
+    constexpr const uint32_t epoch_month_time = 60 * 60 * 24 * 30; // seconds * minutes * hours * days
```
Comment (bertptrs) [inline src/extend_key_expiry/public_key.cpp:10]: People may be confused with the results of their key extension if we take the "all months have 30 days" approach. Isn't there a nice way to say "3 months later" in C++$version?

File: src/extend_key_expiry/public_key.cpp [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+        const auto &hashed_set = _signature.hashed_subpackets();
+
+        /// iterate signature subpackets and retrieve creation and expiration timestamps
+        for (const auto& subpacket : hashed_set) {
+            if (pgp::holds_alternative<pgp::signature_subpacket::signature_creation_time>(subpacket)) {
+                return pgp::get<pgp::signature_subpacket::signature_creation_time>(subpacket).data();
+            }
+        }
+
+        throw std::runtime_error{ "Inexistent creation timestamp subpacket." };
```
Comment (bertptrs) [inline src/extend_key_expiry/public_key.cpp:102]: Nitpick: nonexistent is the more common way to say this. TIL inexistent is a word. Although I'd reword it as "Did not find creation timestamp subpacket."

File: src/generate_derived_key/CMakeLists.txt [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+
+# Add CryptoPP
+target_link_libraries(generate_derived_key CryptoPP::CryptoPP)
+
+# I don't know whether this is the earliest version that supports concepts, but
+# at least 8.2.1 supports it.
+if (CMAKE_COMPILER_IS_GNUCC AND CMAKE_CXX_COMPILER_VERSION VERSION_GREATER 8.2)
+    set_property(TARGET generate_derived_key PROPERTY CXX_STANDARD 20)
+    add_compile_options(-fconcepts -DHAVE_CPP20_CONCEPTS)
+endif()
```
Comment (bertptrs) [inline src/generate_derived_key/CMakeLists.txt:36]: Nitpick: missing newline at EOF.

File: src/generate_derived_key/generate_derived_key.cpp [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+            );
+        }
+
+        // convert the dates to a timestamp
+        std::time_t key_creation_timestamp          = time_utils::tm_to_utc_unix_timestamp(*options.key_creation);
+        std::time_t signature_creation_timestamp    = time_utils::tm_to_utc_unix_timestamp(*options.signature_creation);
+        std::time_t signature_expiration_timestamp  = time_utils::tm_to_utc_unix_timestamp(*options.signature_expiration);
+
+        // select the function with which to generate the packets
+        std::function<std::vector<pgp::packet>(const master_key&, std::string, uint32_t, uint32_t, uint32_t, boost::string_view, bool)> generation_function;
```
Comment (bertptrs) [inline src/generate_derived_key/generate_derived_key.cpp:104]: The answer is probably no, but is there any way to make this type less terrible?

File: src/shared/util/options.cpp [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+            result = false;
+        } else {
+            // no parse, set the fail bit
+            stream.setstate(std::ios_base::failbit);
+        }
+
+        // return the stream for chaining
+        return stream;
+    }
+}
```
Comment (bertptrs) [inline src/shared/util/options.cpp:102]: Nitpick: no newline again.

File: tests/generate.py [hunk 9]
```diff
@@ -31,6 +31,10 @@ def generateDice():
 def generateSymmetricKey():
     return generateString()
 
+def generateExtensionPeriod():
+    options = ['3', '6', '12']
+    return options[random.randint(0, 2)]
```
Comment (bertptrs) [inline tests/generate.py:36]: `return random.choice(options)`

File: tests/generate.py [hunk 10]
```diff
@@ -48,5 +52,6 @@ def generateInput():
         "dice": generateDice(),
         "key": generateSymmetricKey(),
         "context": generateString(minstrlen=8, maxstrlen=8),
-        "key_creation": date_key_creation
+        "key_creation": date_key_creation,
+        "extension_period": generateExtensionPeriod()
```
Comment (bertptrs) [inline tests/generate.py:56]: Please use trailing commas in dictionary literals, and anywhere else for that matter.

File: tests/integration_test.py [hunk 11]
```diff
@@ -130,7 +132,10 @@ def read_line(self, line, timeout_ms = 1000):
                 break
         return line
 
-    def read_all(self):
+    def read_all(self, decode = True):
```
Comment (bertptrs) [inline tests/integration_test.py:135]: PEP8 recommends not putting spaces around default values for arguments.

File: tests/integration_test.py [hunk 12]
```diff
@@ -228,6 +247,18 @@ def import_gpg_packet(filename, **kwargs):
 
     return all(l)
 
+def export_public_key(key_fingerprint, workdir, **kwargs):
+    keyfile = os.path.join(workdir, safe_temporary_name())
+
+    with GPGApplication(["--export", key_fingerprint], also_stderr=True, **kwargs) as app:
+        output = app.read_all(decode = False)
```
Comment (bertptrs) [inline tests/integration_test.py:254]: Why name this argument in particular?

File: tests/integration_test.py [hunk 13]
```diff
@@ -410,7 +455,49 @@ def report_error(appinput, keyfile, rec_seed):
...
 
-def run_test(exec_name, key_class, language_idx = None):
+# Compare the original and the extended key packets
+def compare_extended_key(original, extended, extension_period):
+    # Make a list of the differences between the original and the extended key
+    diffs = [(i, j) for i, j in zip(original, extended) if i != j]
+    for i, j in diffs:
+        # keys should be equal, only some subpackets of the signature change change
+        if not isinstance(i, SignaturePacket) or not isinstance(j, SignaturePacket):
+            raise ValueError("WRONG WRONG")
```
Comment (bertptrs) [inline tests/integration_test.py:465]: I understand the sentiment but this is not a helpful message.

File: tests/integration_test.py [hunk 14]
```diff
@@ -492,31 +579,41 @@ def run_test(exec_name, key_class, language_idx = None):
...
 def main():
     parser = argparse.ArgumentParser()
     parser.add_argument("exec_name", help="generate_derived_key executable")
-    parser.add_argument("-n", "--num_tests", required=False, type=int, default=1,
-                        help="Number of repetitions to run for the tests.")
+    parser.add_argument("exec_expiry_name", help="generate_derived_key executable")
 
     args = parser.parse_args()
     exec_name = args.exec_name
-    num_tests = args.num_tests
```
Comment (bertptrs) [inline tests/integration_test.py:510]: Was this not used?

File: tests/integration_test.py [hunk 15]
```diff
@@ -301,6 +332,20 @@ def generate_initial_key(workdir, exec_name, appinput, language_idx):
...
 
+def extend_key_expiry(workdir, exec_name, public_key_file, appinput, recovery_seed, language_idx):
+    keyfile = os.path.join(workdir, safe_temporary_name())
+    with ExtendExpiryApplication(exec_name, public_key_file, keyfile, appinput) as app:
+        app.write_line(recovery_seed)
+        app.write_line(str(language_idx))
+        app.write_line(appinput.key)
+        app.read_all()
+
+        #if text:
```
Comment (vicsn) [inline tests/integration_test.py:343]: Can this be removed?

File: src/shared/master_key.h [hunk 16]
```diff
@@ -148,7 +148,7 @@ class master_key : public pgp::secure_object<std::array<uint8_t, crypto_kdf_KEYB
             // or until we have processed a valid recovery
             while (std::cin && !recovery_seed.empty()) {
                 // don't have a valid recovery seed yet
-                std::cout << "Enter recovery seed, or press enter to generate a new key: ";
+                std::cout << "Recovery seed: ";
```
Comment (vicsn) [inline src/shared/master_key.h:151]: I think the original version was better and more accurate!

File: src/extend_key_expiry/public_key.cpp (hunk 4 above)
Comment (bysnack) [reply src/extend_key_expiry/public_key.cpp]: We changed it to use days instead of months, it's up to the user to decide how many days the key should be extended

File: src/extend_key_expiry/CMakeLists.txt [hunk 17]
```diff
@@ -1,30 +1,12 @@
...
-)
+    options.cpp)
 
 # Target definition
 add_executable(extend_key_expiry ${extend-source-list})
 set_property(TARGET extend_key_expiry PROPERTY CXX_STANDARD 17)
 
-# Include shared headers
+# Add shared
 target_include_directories(extend_key_expiry PRIVATE ../shared)
```
Comment (bertptrs) [inline src/extend_key_expiry/CMakeLists.txt:11]: Now that `shared` is a proper CMake target, shouldn't this be handled by `target_link_libraries`?

File: src/extend_key_expiry/public_key.cpp [hunk 18]
```diff
@@ -7,7 +7,7 @@
 
 
 namespace {
-    constexpr const uint32_t epoch_month_time = 60 * 60 * 24 * 30; // seconds * minutes * hours * days
+    constexpr const uint32_t epoch_day_time = 60 * 60 * 24; // seconds in a day
```
Comment (bertptrs) [inline src/extend_key_expiry/public_key.cpp:10]: I should've caught this the first time around but `epoch_day_time` and `expiration_timestamp` should both be 64 bit integers for the coming 2k38 apocalypse. I know they're unsigned so we have 68 more years technically, but time is 64 bit in `$current_year`.

File: src/extend_key_expiry/public_key.cpp [hunk 19]
```diff
@@ -152,10 +152,13 @@ namespace key_expiry {
                 switch (rsa_bit_size) {
                     case 2048:
                         retval = generate_key<parameters::rsa<2048>>(master, user_id(), creation_timestamp(), signature_creation_timestamp(), expiration, kdf_context, debug_dump_keys);
+                        break;
```
Comment (bertptrs) [inline src/extend_key_expiry/public_key.cpp:155]: Shouldn't SA or the compiler have caught this?

File: src/shared/CMakeLists.txt [hunk 20]
```diff
@@ -0,0 +1,17 @@
...
+    parameters_ecdsa.cpp
+    parameters_eddsa.cpp
+    parameters_rsa.cpp
+    packet_utils.cpp
+    time_utils.cpp
+    util/options.cpp)
+
+add_library(shared STATIC ${shared-sources})
+
+target_link_libraries(shared
```
Comment (bertptrs) [inline src/shared/CMakeLists.txt:14]: If you link these libraries publicly, you don't have to repeat them in the two binaries.

File: tests/integration_test.py [hunk 21]
```diff
@@ -487,13 +501,14 @@ def compare_extended_key(original, extended, extension_period):
             # get the difference between the extended key and the original key expiry periods
             result = [int(expiration_dates[4 + i]) - int(expiration_dates[i]) for i in range(len(expiration_dates) // 2)]
             
-            # the key got extended for 3, 6 or 12 months and the expiration period is formatted as "XXyXXdXXhXXm" where XX means
+            # the key got extended for X days and the expiration period is formatted as "XXyXXdXXhXXm" where XX means
             # a two digit number and its translated in our array as [y,d,h,m] therefore the previous substraction should be
-            # equivalent to period * 30 days represented as [0,period*30,0,0] but because more than 365 days are actually a year
-            # we need to take this variable into the equation, this can be done with the next formula: period * 30 = d - 365 * y
+            # equivalent to the period of days represented as [years,months,0,0] but because more than 365 days are actually a year
+            # we need to take this variable into the equation, thusu the extension period can be calculated as 365 * y + d
```
Comment (bertptrs) [inline tests/integration_test.py:507]: Typo: `thus`

File: src/shared/CMakeLists.txt (hunk 20 above)
Comment (bysnack) [reply src/shared/CMakeLists.txt:14]: Don't they get linked publicly by default?, I'm not really linking them again in the binaries

File: src/extend_key_expiry/public_key.cpp (hunk 18 above)
Comment (bysnack) [reply src/extend_key_expiry/public_key.cpp:10]: As discussed every timestamp is defined as a 32 bits unsigned integer in the tool and the library, changing is work for another PR

File: src/extend_key_expiry/public_key.cpp (hunk 19 above)
Comment (bysnack) [reply src/extend_key_expiry/public_key.cpp:155]: I think it didn't because I was compiling with clang6 to see if the minimum required version worked, compiling with clang9 showed me a new warning so maybe it would have caught this too.

File: src/extend_key_expiry/public_key.cpp (hunk 18 above)
Comment (vicsn) [reply src/extend_key_expiry/public_key.cpp:10]: @bysnack please make an issue
